File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    public TreeNode invert(TreeNode temp)
    {
        if(temp==null)
        {
            return null;
        }
         TreeNode left = invert(temp.left);
         TreeNode right = invert(temp.right); 
         temp.right=left;
         temp.left=right;

         return temp;
    }

    public TreeNode invertTree(TreeNode root) {
        return invert(root);
        }
    }

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        
        def checkSubTree(root,subRoot) :
            if subRoot== None and root ==None:
                return True
            
            if  subRoot and  root and subRoot.val!=root.val:
                return False
            
            if subRoot==None or root==None:
                return False
            left=checkSubTree(root.left,subRoot.left)
            right=checkSubTree(root.right,subRoot.right)

            return left and right

        def traverse(root,subRoot) :
            if root==None:
                return False
            
            if root.val==subRoot.val :
                result=checkSubTree(root,subRoot)
                if result :
                    return True
                
            left=traverse(root.left,subRoot)
            right=traverse(root.right,subRoot)

            return left or right

        return traverse(root,subRoot)
    
